include/libopencm3/stm32/common/sdio_common_all.h

```diff
@@ -0,0 +1,266 @@
+	BLOCK_LENGTH_4_B     = (uint32_t)(0x2 << SDIO_DCTRL_DBLOCKSIZE_SHIFT),
+	BLOCK_LENGTH_8_B     = (uint32_t)(0x3 << SDIO_DCTRL_DBLOCKSIZE_SHIFT),
+	BLOCK_LENGTH_16_B    = (uint32_t)(0x4 << SDIO_DCTRL_DBLOCKSIZE_SHIFT),
+	BLOCK_LENGTH_32_B    = (uint32_t)(0x5 << SDIO_DCTRL_DBLOCKSIZE_SHIFT),
+	BLOCK_LENGTH_64_B    = (uint32_t)(0x6 << SDIO_DCTRL_DBLOCKSIZE_SHIFT),
+	BLOCK_LENGTH_128_B   = (uint32_t)(0x7 << SDIO_DCTRL_DBLOCKSIZE_SHIFT),
+	BLOCK_LENGTH_256_B   = (uint32_t)(0x8 << SDIO_DCTRL_DBLOCKSIZE_SHIFT),
+	BLOCK_LENGTH_512_B   = (uint32_t)(0x9 << SDIO_DCTRL_DBLOCKSIZE_SHIFT),
+	BLOCK_LENGTH_1024_B  = (uint32_t)(0xA << SDIO_DCTRL_DBLOCKSIZE_SHIFT),
+	BLOCK_LENGTH_2048_B  = (uint32_t)(0xB << SDIO_DCTRL_DBLOCKSIZE_SHIFT),
+	BLOCK_LENGTH_4096_B  = (uint32_t)(0xC << SDIO_DCTRL_DBLOCKSIZE_SHIFT),
+	BLOCK_LENGTH_8192_B  = (uint32_t)(0xD << SDIO_DCTRL_DBLOCKSIZE_SHIFT),
+	BLOCK_LENGTH_16384_B = (uint32_t)(0xE << SDIO_DCTRL_DBLOCKSIZE_SHIFT),
+}SDIO_BlockSize;
+
+typedef enum 
+{
+	SDIO_CEATAEND	= (uint32_t)(1 << 23), //CE-ATA command completion signal received for CMD61
+	SDIO_SDIOIT		= (uint32_t)(1 << 22), //SD I/O received
+	SDIO_RXDAVL		= (uint32_t)(1 << 21), //Data available in receive FIFO
+	SDIO_TXDAVL		= (uint32_t)(1 << 20), //Data available in transmit FIFO 
+	SDIO_RXFIFOE	= (uint32_t)(1 << 19), //Receive FIFO empty 
+	SDIO_TXFIFOE	= (uint32_t)(1 << 18), //Transmit FIFO empty 
+	SDIO_RXFIFOF	= (uint32_t)(1 << 17), //Receive FIFO full 
+	SDIO_TXFIFOF	= (uint32_t)(1 << 16), //Transmit FIFO full 
+	SDIO_RXFIFOHF	= (uint32_t)(1 << 15), //Receive FIFO Half Full 
+	SDIO_TXFIFOHE	= (uint32_t)(1 << 14), //Transmit FIFO Half Empty 
+	SDIO_RXACT		= (uint32_t)(1 << 13), //Data receive in progress 
+	SDIO_TXACT		= (uint32_t)(1 << 12), //Data transmit in progress 
+	SDIO_CMDACT		= (uint32_t)(1 << 11), //Command transfer in progress 
+	SDIO_DBCKEND	= (uint32_t)(1 << 10), //Data block sent/received (CRC check passed) 
+	SDIO_STBITERR	= (uint32_t)(1 << 9),  //Start bit not detected on all data signals in wide bus mode 
+	SDIO_DATAEND	= (uint32_t)(1 << 8),  //Data end (data counter, SDIDCOUNT, is zero) 
+	SDIO_CMDSENT	= (uint32_t)(1 << 7),  //Command sent (no response required) 
+	SDIO_CMDREND	= (uint32_t)(1 << 6),  //Command response received (CRC check passed) 
+	SDIO_RXOVERR	= (uint32_t)(1 << 5),  //Received FIFO overrun error 
+	SDIO_TXUNDERR	= (uint32_t)(1 << 4),  //Transmit FIFO underrun error 
+	SDIO_DTIMEOUT	= (uint32_t)(1 << 3),  //Data timeout 
+	SDIO_CTIMEOUT	= (uint32_t)(1 << 2),  //Command response timeout 
+	SDIO_DCRCFAIL	= (uint32_t)(1 << 1),  //Data block sent/received (CRC check failed) 
+	SDIO_CCRCFAIL	= (uint32_t)(1 << 0),  //Command response received (CRC check failed) 
+}SDIO_Flag;
+
+typedef enum 
+{
+	READ_WAIT_DATA2,
+	READ_WAIT_CLK,
+}SDIO_ReadWaitMode;
+
+typedef struct
+{
+	SDIO_CK_Polarity polarity;
+	bool bypass;
+	bool power_save;
+	SDIO_WideBusMode bus_wide;
+	bool flow_control;
+	uint8_t clock_divider;
+} SDIO_config;
```

Comment:
_definitely_ no typedefs for structs.  Also, the only place we've been using structs for api arguments so far has been the rcc clock structures. I'm not really in favour of mixing the styles, regardless of the benefits of one approach or another.